File: dsa/02-blind-75/arrays-hashing/09-three-sum/solutions/typescript/threeSumTwoPointers.ts
export function threeSum(numbers: number[]): number[][] {
  const sortedNumbers = [...numbers].sort((left, right) => left - right);
  const triplets: number[][] = [];

  for (let anchor = 0; anchor < sortedNumbers.length - 2; anchor += 1) {
    if (anchor > 0 && sortedNumbers[anchor] === sortedNumbers[anchor - 1]) continue;
    if (sortedNumbers[anchor] > 0) break;

    let left = anchor + 1;
    let right = sortedNumbers.length - 1;
    while (left < right) {
      const total = sortedNumbers[anchor] + sortedNumbers[left] + sortedNumbers[right];
      if (total < 0) {
        left += 1;
      } else if (total > 0) {
        right -= 1;
      } else {
        triplets.push([sortedNumbers[anchor], sortedNumbers[left], sortedNumbers[right]]);
        left += 1;
        right -= 1;
        while (left < right && sortedNumbers[left] === sortedNumbers[left - 1]) left += 1;
        while (left < right && sortedNumbers[right] === sortedNumbers[right + 1]) right -= 1;
      }
    }
  }

  return triplets;
}

File: dsa/02-blind-75/arrays-hashing/09-three-sum/solutions/python/three_sum_two_pointers.py
def three_sum(numbers: list[int]) -> list[list[int]]:
    sorted_numbers = sorted(numbers)
    triplets: list[list[int]] = []

    for anchor, value in enumerate(sorted_numbers[:-2]):
        if anchor > 0 and value == sorted_numbers[anchor - 1]:
            continue
        if value > 0:
            break

        left, right = anchor + 1, len(sorted_numbers) - 1
        while left < right:
            total = value + sorted_numbers[left] + sorted_numbers[right]
            if total < 0:
                left += 1
            elif total > 0:
                right -= 1
            else:
                triplets.append([value, sorted_numbers[left], sorted_numbers[right]])
                left += 1
                right -= 1
                while left < right and sorted_numbers[left] == sorted_numbers[left - 1]:
                    left += 1
                while left < right and sorted_numbers[right] == sorted_numbers[right + 1]:
                    right -= 1

    return triplets
